postgres/db.go

```diff
@@ -0,0 +1,701 @@
+// Delete archives or soft deletes the database record for value.
+func (db *DB) Delete(value any) error {
+	if db.db.Error != nil {
+		return db.db.Error
+	}
+
+	res := db.db.Delete(value)
+	if errors.Is(res.Error, schema.ErrUnsupportedDataType) {
+		return fmt.Errorf("%w: cannot parse table name from %T", trails.ErrMissingData, value)
+	}
+
+	if res.Error != nil {
+		return fmt.Errorf("%w: failed deleting %T: %s", trails.ErrUnexpected, value, res.Error)
+	}
+
+	if res.RowsAffected == 0 {
+		return fmt.Errorf("%w: %T", trails.ErrNotFound, value)
+	}
+
+	return nil
+}
+
+// Exec executes SQL query sql, passing values to it.
+//
+// If the query executed does not affect any records, Exec return ErrNotFound.
+// There are many use cases where the caller out to specifically ignore this error,
+// since the execution may not change existing records.
+//
+// Exec does not write any data resulting from the query into Go values.
+func (db *DB) Exec(sql string, values ...any) error {
+	if db.db.Error != nil {
+		return db.db.Error
+	}
+
+	var err error
+	values, err = unwrap(values...)
+	if err != nil && !errors.Is(err, errNilArg) {
+		return err
+	}
+
+	res := db.db.Exec(sql, values...)
+	if res.Error != nil {
+		return fmt.Errorf("%w: %s", trails.ErrUnexpected, res.Error)
+	}
+
+	if res.RowsAffected == 0 {
+		return fmt.Errorf("%w: exec failed to affect any rows", trails.ErrNotFound)
+	}
+
+	return nil
+}
+
+// Exists asserts whether any record matches the current query.
+func (db *DB) Exists() (bool, error) {
+	if db.db.Error != nil {
+		return false, db.db.Error
+	}
+
+	var exists bool
+	// NOTE(dlk): This is weird and can't explain well why *gorm.DB.Session
+	// is necessary in this instance.
+	// Without it, GORM fails to render the current query as a sub-query.
+	//
+	// FIXME(dlk): Scan produces a SELECT * when a bare SELECT is sufficient,
+	// some hack possible to get that marginal optimization?
+	err := db.db.Raw("SELECT EXISTS(?)", db.db.Session(safeGORMSession)).Scan(&exists).Error
+	if err != nil {
+		err = fmt.Errorf("%w: %s", trails.ErrUnexpected, err)
+		return false, err
+	}
+
+	return exists, nil
+}
+
+// Find retrieves all records matching the current query
+// and stores them in dest.
+//
+// If dest is not a valid type for the table queried,
+// then ErrNotValid returns.
+// If no matches are found, Find returns ErrNotFound.
+func (db *DB) Find(dest any) (err error) {
+	badDest := fmt.Errorf("%w: %T cannot be scanned into", trails.ErrNotValid, dest)
+	defer func() {
+		if r := recover(); r != nil {
+			err = badDest
+		}
+	}()
+
+	if db.db.Error != nil {
+		return db.db.Error
+	}
+
+	res := db.db.Find(dest)
+	err = res.Error
+	if err != nil && errSQLScan.MatchString(err.Error()) {
+		return badDest
+	}
+
+	if err != nil && errSQLSyntax.MatchString(err.Error()) {
+		return fmt.Errorf("%w: %s", trails.ErrNotValid, err)
+	}
+
+	if err != nil {
+		return fmt.Errorf("%w: %s", trails.ErrUnexpected, err)
+	}
+
+	if res.RowsAffected == 0 {
```

Comment:
This one surprised me. Typically, `.First` is where we are worried about something not being found and where gorm returns an error. But `.Find` can return zero results and this is not an error state. Is this intentional? I see the extra handling in portal `if err != nil && !errors.Is(err, trails.ErrNotFound) {` so assuming it is intentional. This is going to be hard for me to get used to given that it different from current functionality. What's the strategy here that makes it worth changing?

Reply:
Doing some digging, it appears that `len(data) == 0` being something to handle is more the exception than the rule and it actually changes the logic in some places where the extra handling is now applied when it probably shouldn't be. e.g. groups#ExtendMeetings

Reply:
Right, there'll be more `!errors.Is(err, trails.ErrNotFound)`. The reasoning is this two-fold: (1) eliminates a lot of `if len(data) == 0` code blocks; and (2) makes `First` and `Find` have identical behavior instead of one having this additional check and the other not.

Definitely call out any cases where it looks

Reply:
Copying from chatting with @steven-klein:

Every Find ought to be followed up with at least `if err != nil`, so you're halfway there to ignore `ErrNotFound` if you so choose. Same guidance applies to `Update` and `Exec`. I understand other ORMs that GORM copies from don't have the error-first approach that Go does and so it's a bit foreign.

I'll grant the empty result set isn't an error in a strict sense. The query ran and nothing returned. No issues there. But, having `postgres.DB` wrap up the state in an error enables simpler logic on caller's side given our conventions of handling errors.

To reduce it down, the gist of my stance is: "hey it isn't annoying you have to check for errors and `if len(data) != 0`?!? What if you could do that in one line?"

Reply:
I don't really see this as an argument around go being an error first approach, but that most ORMs only error when the API semantics say “exactly one row must exist”. This is the case with `.First()` but not `.Find()` or `.Delete` etc, and I imagine devs coming from sqlalchemy, django orm, Sequelize, activerecord, etc.. would be caught out by this choice. If we want this type of path, I would update the API to have `UpdateOne()`, `DeleteOne()`, etc.. Let's let it percolate over the weekend and we can make a decision monday morning on whether to keep this or revert to the existing behaviour.